notebooks/community/bigquery_ml/Anomaly_detection_in_Cloud_Audit_logs_with_BQML.ipynb

```diff
@@ -0,0 +1,1104 @@
+        ")\n",
+        "RETURNS STRING\n",
+        "AS (\n",
+        "  CASE\n",
+        "    WHEN caller_supplied_user_agent LIKE \"Mozilla/%\" THEN 'Cloud Console'\n",
+        "    WHEN caller_supplied_user_agent LIKE \"google-cloud-sdk gcloud/%\" THEN 'gcloud CLI'\n",
+        "    WHEN caller_supplied_user_agent LIKE \"google-api-go-client/% Terraform/%\" THEN 'Terraform'\n",
+        "    ELSE 'other'\n",
+        "  END\n",
+        ");\"\"\".format(\n",
+        "    PROJECT_ID, BQ_DATASET_NAME, UDF_NAME\n",
+        ")\n",
+        "\n",
+        "execute_sql(sql)\n",
+        "print(f\"Created UDF {PROJECT_ID}.{BQ_DATASET_NAME}.{UDF_NAME}\")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "BciIe3NEGn7c"
+      },
+      "source": [
+        "Query the log source to extract the training data with fields of interest"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "pO3BxDg8rKWx"
+      },
+      "outputs": [],
+      "source": [
+        "# Query to extract training data with fields of interest\n",
+        "query_str = \"\"\" SELECT\n",
+        "    EXTRACT(DATE FROM timestamp) AS day,\n",
+        "    IFNULL(proto_payload.audit_log.authentication_info.principal_email, \"unknown\") as principal_email,\n",
+        "    IFNULL(proto_payload.audit_log.method_name, \"unknown\") as action,\n",
+        "    IFNULL(resource.type, \"unknown\") as resource_type,\n",
+        "    {3}.getResourceId(resource.type, resource.labels) AS resource_id,\n",
+        "    -- proto_payload.audit_log.resource_name as resource_name,\n",
+        "    SPLIT(log_name, '/')[SAFE_OFFSET(0)] as container_type,\n",
+        "    SPLIT(log_name, '/')[SAFE_OFFSET(1)] as container_id,\n",
+        "    {3}.getChannelType(proto_payload.audit_log.request_metadata.caller_supplied_user_agent) AS channel,\n",
+        "    IFNULL(proto_payload.audit_log.request_metadata.caller_ip, \"unknown\") as ip,\n",
+        "    COUNT(*) counter,\n",
+        "    -- ANY_VALUE(resource) as resource,           -- for debugging\n",
+        "    -- ANY_VALUE(proto_payload) as proto_payload  -- for debugging\n",
+        "  FROM  `{0}.{1}.{2}`\n",
+        "  WHERE\n",
+        "    -- log_id = \"cloudaudit.googleapis.com/activity\" AND\n",
+        "    timestamp > TIMESTAMP_SUB(CURRENT_TIMESTAMP(), INTERVAL 360 DAY)\n",
+        "  GROUP BY\n",
+        "    day, principal_email, action, resource_type, resource_id, container_type, container_id, channel, ip, log_name\n",
+        "  ORDER BY\n",
+        "    day DESC, principal_email, action\"\"\".format(\n",
+        "    logSourceProject, logSourceBqDataset, logSourceBqTable, BQ_DATASET_NAME\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "wAOZwnhOG2Jf"
+      },
+      "source": [
+        "Lets view the training data dataframe"
```

Comment:
Modify to use second person and an imperative style (Example: "Let's view the training data dataframe" -> "View the training data dataframe")